File: 2461-amount-of-time-for-binary-tree-to-be-infected/amount-of-time-for-binary-tree-to-be-infected.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int AmountOfTime(TreeNode root, int start) {
        var parents = new Dictionary<int, TreeNode>();
        var visited = new Dictionary<int, bool>();
        
        var q1 = new Queue<TreeNode>();
        q1.Enqueue(root);

        TreeNode startNode = null;

        while (q1.Any()) {
            var front = q1.Dequeue();
            visited.Add(front.val, false);
            if (front.left != null) {
                parents.Add(front.left.val, front);
                q1.Enqueue(front.left);
            }
            if (front.right != null) {
                parents.Add(front.right.val, front);
                q1.Enqueue(front.right);
            }
            if (front.val == start) {
                startNode = front;
            }
        }

        var time = 0;
        var q = new Queue<(TreeNode, int)>();
        q.Enqueue((startNode, 0));
        visited[start] = true;

        while (q.Any()) {
            var (front, dist) = q.Dequeue();
            time = dist;

            if (front.left != null && !visited[front.left.val]) {
                visited[front.left.val] = true;
                q.Enqueue((front.left, dist+1));
            }
            if (front.right != null && !visited[front.right.val]) {
                visited[front.right.val] = true;
                q.Enqueue((front.right, dist+1));
            }
            if (parents.ContainsKey(front.val) && !visited[parents[front.val].val]) {
                visited[parents[front.val].val] = true;
                q.Enqueue((parents[front.val], dist+1));
            }
        }

        return time;
    }
}

File: 2461-amount-of-time-for-binary-tree-to-be-infected/amount-of-time-for-binary-tree-to-be-infected.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int depthFromTarget = 0;
public:
    int amountOfTime(TreeNode* root, int start) {
        unordered_map<int, TreeNode*> parents;
        unordered_map<int, bool> visited;
        queue<TreeNode*> q1;
        TreeNode* startNode;

        q1.push(root);

        while (!q1.empty()) {
            TreeNode* front = q1.front();
            q1.pop();

            if (front->val == start) {
                startNode = front;
            }
            if (front->left != NULL) {
                parents[front->left->val] = front;
                q1.push(front->left);
            }
            if (front->right != NULL) {
                parents[front->right->val] = front;
                q1.push(front->right);
            }
        }

        queue<pair<TreeNode*, int>> q2;

        q2.push({startNode, 0});
        visited[start] = true;
        int time = 0;

        while (!q2.empty()) {
            TreeNode* node = q2.front().first;
            int dist = q2.front().second;
            q2.pop();

            time = dist;

            if (node->left != NULL && !visited[node->left->val]) {
                q2.push({node->left, dist+1});
                visited[node->left->val] = true;
            }
            if (node->right != NULL && !visited[node->right->val]) {
                q2.push({node->right, dist+1});
                visited[node->right->val] = true;
            }
            if (parents[node->val] != NULL && !visited[parents[node->val]->val]) {
                q2.push({parents[node->val], dist+1});
                visited[parents[node->val]->val] = true;
            }
        }

        return time;
    }
};